### DISCUSSION TITLE
Can I add plugins to package.json?

### DISCUSSION THREAD

**Discussion by tking16** [appium/appium#20408] [Q&A]
I have plugins in my `package.json` file but they don't get installed when I run `npm install`. My file looks as follows:

```json
{
  "name": "app-at",
  "version": "1.0.0",
  "description": "at_app ======",
  "main": "index.js",
  "directories": {
    "test": "tests"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "mjpeg-consumer": "^2.0.0",
    "npm-bin": "^0.1.0",
    "unzipper": "^0.10.14",
    "wd": "^1.14.0"
  },
  "devDependencies": {
    "appium-dashboard": "^2.0.2",
    "appium-device-farm": "^8.2.3",
    "appium-uiautomator2-driver": "^2.43.4",
    "appium-xcuitest-driver": "^7.15.2"
  }
}

```
Namely it's `appium-device-farm` I'm looking to get installed

**Comment by portyu9** [comment]
Hi @tking16 

Yes — Appium drivers/plugins **can** be managed through your project's `package.json`, but there is an important detail missing from your example: **Appium itself also needs to be a dependency of that project**.

For example:

```json
{
  "devDependencies": {
    "appium": "^2.0.0",
    "appium-device-farm": "^8.2.3",
    "appium-uiautomator2-driver": "^2.43.4",
    "appium-xcuitest-driver": "^7.15.2"
  }
}
```

Then run:

```bash
npm install
npx appium --use-plugins=device-farm
```

When Appium is installed locally in the same npm project, `npx appium` can discover compatible drivers/plugins listed in that project's `package.json`.

Your current `package.json` contains the extensions but **does not contain `appium` itself**, so if you are starting a globally installed `appium`, that Appium installation will normally use extensions managed under its `APPIUM_HOME` instead of treating your project's `node_modules` as its installed extensions.

If you want to keep using a global Appium installation, install Device Farm through Appium's extension CLI instead:

```bash
appium plugin install --source=npm appium-device-farm
appium --use-plugins=device-farm
```

You can verify what that Appium installation sees with:

```bash
appium plugin list --installed
```

Also, if by "doesn't get installed" you literally mean `appium-device-farm` is missing from your project's `node_modules` after `npm install`, check:

```bash
npm ls appium-device-farm
```

Since it is under `devDependencies`, npm will omit it if you are installing with production/dev-dependency omission settings such as `npm install --omit=dev`.

One version-related note: your example uses Appium 2-era driver/plugin versions. Current Device Farm releases target newer Appium versions, so keep the Appium core, drivers, and Device Farm versions mutually compatible rather than upgrading only one of them.

Please mark this answer as accepted if it helped, thank you.
